File: scripts/lib/atproto/clients/records.py
from __future__ import annotations

from typing import Any, Optional

from ..transport import TransportLayer


class RecordsClient:
    def __init__(self, transport: TransportLayer):
        self._t = transport

    def create_record(
        self,
        repo: str,
        collection: str,
        record: dict[str, Any],
        token: str,
        rkey: Optional[str] = None,
        validate: bool = True,
    ) -> dict:
        body: dict[str, Any] = {
            "repo": repo,
            "collection": collection,
            "record": record,
            "validate": validate,
        }
        if rkey:
            body["rkey"] = rkey
        return self._t.post("com.atproto.repo.createRecord", body, token=token)

    def get_record(self, repo: str, collection: str, rkey: str) -> dict:
        return self._t.get(
            "com.atproto.repo.getRecord",
            {"repo": repo, "collection": collection, "rkey": rkey},
        )

    def delete_record(self, repo: str, collection: str, rkey: str, token: str) -> None:
        self._t.post(
            "com.atproto.repo.deleteRecord",
            {"repo": repo, "collection": collection, "rkey": rkey},
            token=token,
        )

    def put_record(
        self,
        repo: str,
        collection: str,
        rkey: str,
        record: dict[str, Any],
        token: str,
    ) -> dict:
        return self._t.post(
            "com.atproto.repo.putRecord",
            {"repo": repo, "collection": collection, "rkey": rkey, "record": record},
            token=token,
        )

    def list_records(
        self,
        repo: str,
        collection: str,
        limit: int = 50,
        token: Optional[str] = None,
    ) -> dict:
        return self._t.get(
            "com.atproto.repo.listRecords",
            {"repo": repo, "collection": collection, "limit": limit},
            token=token,
        )

    def apply_writes(
        self,
        repo: str,
        writes: list[dict[str, Any]],
        token: str,
    ) -> dict:
        return self._t.post(
            "com.atproto.repo.applyWrites",
            {"repo": repo, "writes": writes},
            token=token,
        )

File: scripts/lib/deno/clients/records.ts
import { TransportLayer } from "../transport.ts";

export class RecordsClient {
  constructor(private transport: TransportLayer) {}

  async createRecord(
    repo: string,
    collection: string,
    record: Record<string, any>,
    token: string,
    options: { rkey?: string; validate?: boolean } = {}
  ) {
    const body: Record<string, any> = {
      repo,
      collection,
      record,
      validate: options.validate ?? true,
    };
    if (options.rkey) body.rkey = options.rkey;
    return await this.transport.post("com.atproto.repo.createRecord", body, token);
  }

  async getRecord(repo: string, collection: string, rkey: string) {
    return await this.transport.get("com.atproto.repo.getRecord", {
      repo,
      collection,
      rkey,
    });
  }

  async deleteRecord(repo: string, collection: string, rkey: string, token: string) {
    return await this.transport.post(
      "com.atproto.repo.deleteRecord",
      { repo, collection, rkey },
      token
    );
  }

  async putRecord(
    repo: string,
    collection: string,
    rkey: string,
    record: Record<string, any>,
    token: string
  ) {
    return await this.transport.post(
      "com.atproto.repo.putRecord",
      { repo, collection, rkey, record },
      token
    );
  }

  async listRecords(
    repo: string,
    collection: string,
    options: { limit?: number; token?: string } = {}
  ) {
    return await this.transport.get(
      "com.atproto.repo.listRecords",
      { repo, collection, limit: options.limit ?? 50 },
      options.token
    );
  }

  async applyWrites(repo: string, writes: Array<Record<string, any>>, token: string) {
    return await this.transport.post(
      "com.atproto.repo.applyWrites",
      { repo, writes },
      token
    );
  }
}
